File: MelodyCommons__backend/api/playlists.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List

from database import get_db
import crud
import schemas
from auth import get_current_user

router = APIRouter(prefix="/playlists", tags=["playlists"])


@router.get("/", response_model=List[schemas.Playlist])
def get_playlists(
        db: Session = Depends(get_db),
        current_user: schemas.User = Depends(get_current_user)
):
    """获取所有歌单"""
    playlists = crud.get_playlists(db)
    return playlists


@router.get("/{playlist_id}", response_model=schemas.Playlist)
def get_playlist(
        playlist_id: int,
        db: Session = Depends(get_db),
        current_user: schemas.User = Depends(get_current_user)
):
    """获取单个歌单详情"""
    playlist = crud.get_playlist(db, playlist_id=playlist_id)
    if playlist is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Playlist not found"
        )
    return playlist


@router.post("/", response_model=schemas.Playlist)
def create_playlist(
        playlist: schemas.PlaylistCreate,
        db: Session = Depends(get_db),
        current_user: schemas.User = Depends(get_current_user)
):
    """创建歌单"""
    try:
        return crud.create_playlist(db=db, playlist=playlist)
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to create playlist"
        )


@router.put("/{playlist_id}", response_model=schemas.Playlist)
def update_playlist(
        playlist_id: int,
        playlist_update: schemas.PlaylistUpdate,
        db: Session = Depends(get_db),
        current_user: schemas.User = Depends(get_current_user)
):
    """更新歌单信息"""
    playlist = crud.get_playlist(db, playlist_id=playlist_id)
    if playlist is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Playlist not found"
        )

    updated_playlist = crud.update_playlist(db, playlist_id=playlist_id, playlist_update=playlist_update)
    return updated_playlist


@router.delete("/{playlist_id}")
def delete_playlist(
        playlist_id: int,
        db: Session = Depends(get_db),
        current_user: schemas.User = Depends(get_current_user)
):
    """删除歌单"""
    playlist = crud.get_playlist(db, playlist_id=playlist_id)
    if playlist is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Playlist not found"
        )

    success = crud.delete_playlist(db, playlist_id=playlist_id)
    if success:
        return {"message": "Playlist deleted successfully"}
    else:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to delete playlist"
        )


@router.post("/{playlist_id}/songs/{song_id}")
def add_song_to_playlist(
        playlist_id: int,
        song_id: int,
        order_index: int = None,
        db: Session = Depends(get_db),
        current_user: schemas.User = Depends(get_current_user)
):
    """添加歌曲到歌单"""
    # 检查歌单是否存在
    playlist = crud.get_playlist(db, playlist_id=playlist_id)
    if playlist is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Playlist not found"
        )

    # 检查歌曲是否存在
    song = crud.get_song(db, song_id=song_id)
    if song is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Song not found"
        )

    # 添加歌曲到歌单
    result = crud.add_song_to_playlist(db, playlist_id=playlist_id, song_id=song_id, order_index=order_index)
    if result is None:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="Song already exists in playlist"
        )

    return {"message": "Song added to playlist successfully"}


@router.delete("/{playlist_id}/songs/{song_id}")
def remove_song_from_playlist(
        playlist_id: int,
        song_id: int,
        db: Session = Depends(get_db),
        current_user: schemas.User = Depends(get_current_user)
):
    """从歌单移除歌曲"""
    # 检查歌单是否存在
    playlist = crud.get_playlist(db, playlist_id=playlist_id)
    if playlist is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Playlist not found"
        )

    success = crud.remove_song_from_playlist(db, playlist_id=playlist_id, song_id=song_id)
    if success:
        return {"message": "Song removed from playlist successfully"}
    else:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Song not found in playlist"
        )


@router.get("/{playlist_id}/songs", response_model=List[schemas.SongInPlaylist])
def get_playlist_songs(
        playlist_id: int,
        db: Session = Depends(get_db),
        current_user: schemas.User = Depends(get_current_user)
):
    """获取歌单内歌曲（按顺序）"""
    # 检查歌单是否存在
    playlist = crud.get_playlist(db, playlist_id=playlist_id)
    if playlist is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Playlist not found"
        )

    playlist_songs = crud.get_playlist_songs(db, playlist_id=playlist_id)
    return [
        schemas.SongInPlaylist(
            song=ps.song,
            order_index=ps.order_index,
            added_at=ps.added_at
        )
        for ps in playlist_songs
    ]


@router.put("/{playlist_id}/songs/order")
def update_playlist_song_order(
        playlist_id: int,
        order_data: schemas.PlaylistSongOrder,
        db: Session = Depends(get_db),
        current_user: schemas.User = Depends(get_current_user)
):
    """调整歌单内歌曲顺序"""
    # 检查歌单是否存在
    playlist = crud.get_playlist(db, playlist_id=playlist_id)
    if playlist is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Playlist not found"
        )

    try:
        success = crud.update_playlist_song_order(db, playlist_id=playlist_id, song_orders=order_data.song_orders)
        if success:
            return {"message": "Playlist song order updated successfully"}
        else:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to update playlist song order"
            )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to update playlist song order: {str(e)}"
        )

File: melodycommons__frontend/src/stores/playlists.ts
import {defineStore} from 'pinia'
import {ref} from 'vue'
import {playlistsApi} from '@/api/playlists'
import type {Playlist, PlaylistCreate, PlaylistUpdate, SongInPlaylist} from '@/types'

export const usePlaylistsStore = defineStore('playlists', () => {
    const playlists = ref<Playlist[]>([])
    const currentPlaylist = ref<Playlist | null>(null)
    const loading = ref(false)

    // 加载歌单列表
    const loadPlaylists = async () => {
        loading.value = true
        try {
            playlists.value = await playlistsApi.getPlaylists()
        } catch (error) {
            console.error('Failed to load playlists:', error)
            throw error
        } finally {
            loading.value = false
        }
    }

    // 获取单个歌单
    const getPlaylist = async (id: number): Promise<Playlist> => {
        try {
            const playlist = await playlistsApi.getPlaylist(id)
            currentPlaylist.value = playlist
            return playlist
        } catch (error) {
            console.error('Failed to get playlist:', error)
            throw error
        }
    }

    // 创建歌单
    const createPlaylist = async (data: PlaylistCreate): Promise<Playlist> => {
        try {
            const newPlaylist = await playlistsApi.createPlaylist(data)
            playlists.value.unshift(newPlaylist)
            return newPlaylist
        } catch (error) {
            console.error('Failed to create playlist:', error)
            throw error
        }
    }

    // 更新歌单
    const updatePlaylist = async (id: number, data: PlaylistUpdate): Promise<Playlist> => {
        try {
            const updatedPlaylist = await playlistsApi.updatePlaylist(id, data)

            // 更新本地数据
            const index = playlists.value.findIndex(p => p.id === id)
            if (index > -1) {
                playlists.value[index] = updatedPlaylist
            }

            if (currentPlaylist.value?.id === id) {
                currentPlaylist.value = updatedPlaylist
            }

            return updatedPlaylist
        } catch (error) {
            console.error('Failed to update playlist:', error)
            throw error
        }
    }

    // 删除歌单
    const deletePlaylist = async (id: number): Promise<void> => {
        try {
            await playlistsApi.deletePlaylist(id)

            // 从本地数据中移除
            const index = playlists.value.findIndex(p => p.id === id)
            if (index > -1) {
                playlists.value.splice(index, 1)
            }

            if (currentPlaylist.value?.id === id) {
                currentPlaylist.value = null
            }
        } catch (error) {
            console.error('Failed to delete playlist:', error)
            throw error
        }
    }

    // 获取歌单中的歌曲
    const getPlaylistSongs = async (id: number): Promise<SongInPlaylist[]> => {
        try {
            return await playlistsApi.getPlaylistSongs(id)
        } catch (error) {
            console.error('Failed to get playlist songs:', error)
            throw error
        }
    }

    // 添加歌曲到歌单
    const addSongToPlaylist = async (playlistId: number, songId: number): Promise<void> => {
        try {
            await playlistsApi.addSongToPlaylist(playlistId, songId)
        } catch (error) {
            console.error('Failed to add song to playlist:', error)
            throw error
        }
    }

    // 从歌单移除歌曲
    const removeSongFromPlaylist = async (playlistId: number, songId: number): Promise<void> => {
        try {
            await playlistsApi.removeSongFromPlaylist(playlistId, songId)
        } catch (error) {
            console.error('Failed to remove song from playlist:', error)
            throw error
        }
    }

    // 更新歌单中歌曲顺序
    const updatePlaylistSongOrder = async (
        playlistId: number,
        songOrders: Array<{ song_id: number; order_index: number }>
    ): Promise<void> => {
        try {
            await playlistsApi.updatePlaylistSongOrder(playlistId, {song_orders: songOrders})
        } catch (error) {
            console.error('Failed to update playlist song order:', error)
            throw error
        }
    }

    return {
        playlists,
        currentPlaylist,
        loading,
        loadPlaylists,
        getPlaylist,
        createPlaylist,
        updatePlaylist,
        deletePlaylist,
        getPlaylistSongs,
        addSongToPlaylist,
        removeSongFromPlaylist,
        updatePlaylistSongOrder
    }
})